src/librustc_trans/trans/foreign.rs

```diff
@@ -757,7 +769,41 @@ pub fn trans_rust_fn_with_foreign_abi<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
         debug!("calling llrustfn = {}, t = {}",
                ccx.tn().val_to_string(llrustfn), t.repr(ccx.tcx()));
         let attributes = base::get_fn_llvm_attributes(ccx, t);
-        let llrust_ret_val = builder.call(llrustfn, llrust_args.as_slice(), Some(attributes));
+        let llrust_ret_val = if attr::contains_name(attrs, "can_unwind") {
+            debug!("fn can unwind - using call");
+            builder.call(llrustfn, llrust_args.as_slice(), Some(attributes))
+        } else {
+            debug!("fn can't unwind - using invoke");
+            // Create the landing pad and the return blocks
+            let ptr = "catch\0".as_ptr();
+            let catch_block = llvm::LLVMAppendBasicBlockInContext(ccx.llcx(), llwrapfn,
+                                                                ptr as *const _);
+
+            let ptr = "return\0".as_ptr();
+            let return_block = llvm::LLVMAppendBasicBlockInContext(ccx.llcx(), llwrapfn,
+                                                                ptr as *const _);
+
+
+            let llrust_ret_val = builder.invoke(llrustfn, llrust_args.as_slice(),
+                                                return_block, catch_block,
+                                                Some(attributes));
+
+            // Populate the landing pad
+            builder.position_at_end(catch_block);
+            let pad_ty = Type::struct_(ccx, &[Type::i8p(ccx), Type::i32(ccx)], false);
+            let llpersonality = base::get_eh_personality(ccx);
+
+            let pad_ret_val = builder.landing_pad(pad_ty, llpersonality, 1us);
+            builder.add_clause(pad_ret_val, C_null(Type::i8p(ccx)));
+            let trap = ccx.get_intrinsic(&("llvm.trap"));
```

Comment:
Yeah I was thinking that this could take the same strategy as divide-by-zero or out-of-bounds only hardwired to an abort function instead of a panic function. I didn't know about the debugger though, that's pretty slick!